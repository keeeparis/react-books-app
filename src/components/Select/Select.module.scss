.select {
  background-color: var(--color-1);
  user-select: none;
  position: relative;
  width: 250px;
  box-shadow: 1px 6px 20px 1px rgb(0 0 0 / 14%);
  border-radius: 8px;

  &::after {
    content: '';
    position: absolute;
    top: 30%;
    right: 15px;
    width: 10px;
    height: 10px;
    border-bottom: 1px solid black;
    border-left: 1px solid black;
    transform: rotate(-45deg);
    transition: all 0.2s ease;
    pointer-events: none;
  }

  &.rotate::after {
    transform: rotate(-225deg);
    top: 45%;
  }

  &:focus-visible {
    outline: 2px solid var(--color-3);
  }
}

.selectShow {
  cursor: pointer;
  font-size: 1.2em;
  text-transform: capitalize;
  font-weight: 600;
  padding: 10px 0 10px 20px;
  transition: all 0.2s ease;
  border-radius: 8px;

  &:focus-visible {
    outline: 3px solid var(--color-3);
  }

  &:hover {
    background-color: var(--color-1-dark);
  }
}

.selectUl {
  position: absolute;
  width: 100%;
  margin: 10px 0 0 0px;
  background: var(--color-3);
  padding: 0;
  border: 1px solid var(--color-1);
  z-index: 1;
  border-radius: 8px;
}

.selectLi {
  padding: 5px 0px 5px 10px;
  list-style: none;
  cursor: pointer;
  // margin-bottom: 3px;
  color: white;
  position: relative;

  &:first-child {
    border-top-left-radius: 8px;
    border-top-right-radius: 8px;
  }

  &:hover {
    background-color: var(--color-3-dark);
  }

  &:last-child {
    // margin-bottom: 0;
    border-bottom-left-radius: 8px;
    border-bottom-right-radius: 8px;
  }

  &.active {
    background-color: var(--color-1);
    color: black;
    cursor: auto;
  }
}
