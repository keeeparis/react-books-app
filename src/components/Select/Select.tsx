import React, { FC, useState } from 'react'
import styles from './Select.module.scss'
import cn from 'classnames'
import useClickedOutside from '../../hooks/useClickedOutside'

interface SelectType {
  data: string[]
  onClick: (option: string) => void
  value: string
}

const Select: FC<SelectType> = ({ data, onClick, value }) => {
  const [isVisible, setIsVisible] = useState(false)

  const closeOptions = () => {
    setIsVisible(false)
  }

  const { ref } = useClickedOutside(closeOptions)

  // TODO: Разделить на две функциии и прописать типы
  const toggleOptions = (e: any) => {
    if (e.type === 'click') {
      setIsVisible((prev) => !prev)
    } else {
      if (e.keyCode === 13 || e.keyCode === 32) {
        setIsVisible((prev) => !prev)
      }
    }
  }

  const handleClick = (option: string) => (e: any) => {
    if (e.type === 'click') {
      closeOptions()
      onClick(option)
    } else {
      if (e.keyCode === 13 || e.keyCode === 32) {
        closeOptions()
        onClick(option)
      }
    }
  }

  return (
    <div
      className={cn(styles.select, { [styles.rotate]: isVisible })}
      ref={ref}
    >
      <div
        className={styles.selectShow}
        onClick={toggleOptions}
        tabIndex={0}
        onKeyDown={toggleOptions}
      >
        {value}
      </div>
      {isVisible && (
        <ul className={styles.selectUl}>
          {data.map((option) => (
            <li
              className={cn(styles.selectLi, {
                [styles.active]: value === option,
              })}
              onClick={handleClick(option)}
              onKeyDown={handleClick(option)}
              key={option}
              tabIndex={0}
            >
              {option.toUpperCase()}
            </li>
          ))}
        </ul>
      )}
    </div>
  )
}

export default Select
