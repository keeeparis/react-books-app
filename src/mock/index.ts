export const categories = [
  'All',
  'Art',
  'Biography',
  'Computers',
  'History',
  'Medical',
  'Poetry',
  'Fiction',
]

export const sorting = ['relevance', 'newest']
