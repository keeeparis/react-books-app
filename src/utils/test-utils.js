import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom'
import userEvent from '@testing-library/user-event'
import 'regenerator-runtime/runtime'
import { Provider } from 'react-redux'
import { store } from '../redux/store/store'
import { MemoryRouter } from 'react-router-dom'
import '../i18n/config'

const AllTheProviders = ({ children }) => {
  return (
    <Provider store={store}>
      <MemoryRouter>{children}</MemoryRouter>
    </Provider>
  )
}

const customRender = (ui, options) => {
  return render(ui, { wrapper: AllTheProviders, ...options })
}

export * from '@testing-library/react'
export * from '@testing-library/jest-dom'
export { userEvent }
export * from 'regenerator-runtime/runtime'
export * from '../i18n/config'

export { customRender as render }
