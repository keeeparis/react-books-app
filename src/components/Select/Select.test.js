import React from 'react'
import { render, screen, fireEvent, userEvent } from 'test-utils'
import { categories, sorting } from '../../mock'

import Select from './Select'

const props = {
  data: categories,
  value: categories[0],
  onClick: jest.fn(),
}

const props2 = {
  data: sorting,
  value: sorting[0],
  onClick: jest.fn(),
}

it('test Select with categories Component', () => {
  const { asFragment } = render(<Select {...props} />)

  expect(asFragment()).toMatchSnapshot()
})

it('test Select with sorting Component', () => {
  const { asFragment } = render(<Select {...props2} />)

  expect(asFragment()).toMatchSnapshot()
})

beforeEach(() => {
  render(<Select {...props} />)
})

describe('Select', () => {
  test('no items initially', async () => {
    expect(screen.queryAllByRole('listitem')).toHaveLength(0)
  })
  test('after click items are', () => {
    fireEvent.click(screen.getByRole('selected'))
    expect(screen.getAllByRole('listitem')).toHaveLength(categories.length)
  })
  test('register click on item and hide items', async () => {
    fireEvent.click(screen.getByRole('selected'))
    fireEvent.click(screen.getByText(/COMPUTERS/))
    expect(props.onClick).toHaveBeenCalledTimes(1)
    expect(screen.queryAllByRole('listitem')).toHaveLength(0)
  })
  test('register click on selected and hide items', async () => {
    fireEvent.click(screen.getByRole('selected'))
    fireEvent.click(screen.getByRole('selected'))
    expect(screen.queryAllByRole('listitem')).toHaveLength(0)
  })
})
