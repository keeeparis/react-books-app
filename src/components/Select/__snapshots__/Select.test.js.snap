// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Select with categories Component 1`] = `
<DocumentFragment>
  <div
    class="select"
  >
    <div
      class="selectShow"
      role="selected"
      tabindex="0"
    >
      all
    </div>
  </div>
</DocumentFragment>
`;

exports[`test Select with sorting Component 1`] = `
<DocumentFragment>
  <div
    class="select"
  >
    <div
      class="selectShow"
      role="selected"
      tabindex="0"
    >
      relevance
    </div>
  </div>
</DocumentFragment>
`;
